File: contests/leetcode-biweekly182/q2.py
# problem: https://leetcode.cn/contest/biweekly-contest-182/problems/minimum-flips-to-make-binary-string-coherent/
from collections import Counter
from itertools import count
from typing import *

class Solution:
    def minFlips(self, s: str) -> int:
        cnts = Counter(s)
        candidate = 1000000
        if (cnts["1"] == 2 and s[0] == "1" and s[-1] == "1"):
            return 0
        if (cnts["1"] >= 2):
            candidate = cnts["1"] - 2
            firstOne = 0
            secondOne = 0
            for c in s:
                if c == "1":
                    break
                else:
                    firstOne += 1
            for c in reversed(s):
                if c == "1":
                    break
                else:
                    secondOne += 1
            candidate += firstOne + secondOne
        return max(min([cnts["0"], cnts["1"] - 1, candidate]), 0)


def main():
    test = Solution()
    print(test.minFlips("1010"), "e1")
    print(test.minFlips("0110"), "e1")
    print(test.minFlips("1000"), "e0")
    print(test.minFlips("1001"), "e0")
    print(test.minFlips("100011"), "e1")


if __name__ == "__main__":
    main()

#
# Created by: jing At: 2026-05-09 22:37:28
#

File: contests/leetcode-biweekly183/q2.cpp
// problem:
// https://leetcode.cn/contest/biweekly-contest-183/problems/minimum-operations-to-make-array-modulo-alternating-i/
#include <algorithm>
#include <iostream>
#include <ranges>
#include <vector>
#include "../utils/vector_helper.h"


using namespace std;

class Solution {
public:
    int minOperations(vector<int> &nums, int k) {
        for (auto &num: nums) {
            num %= k;
        }
        // show_vector(nums);
        int n = nums.size();
        int mnx1 = INT_MAX, mnx2 = INT_MAX, px1 = 0;
        for (int i = 0; i < k; ++i) {
            int val = 0;
            for (int j = 0; j < n; j += 2) {
                val += min(min(abs(nums[j] - i), abs(nums[j] - (i + k))), nums[j] + k - i);
            }
            if (val <= mnx1) {
                mnx2 = mnx1;
                mnx1 = val;
                px1 = i;
            } else if (val <= mnx2) {
                mnx2 = val;
            }
        }
        int mny1 = INT_MAX, mny2 = INT_MAX, py1 = 0;
        for (int i = 0; i < k; ++i) {
            int val = 0;
            for (int j = 1; j < n; j += 2) {
                val += min(min(abs(nums[j] - i), abs(nums[j] - (i + k))), nums[j] + k - i);
            }
            if (val <= mny1) {
                mny2 = mny1;
                mny1 = val;
                py1 = i;
            } else if (val <= mny2) {
                mny2 = val;
            }
        }
        if (px1 == py1) {
            return min(mnx1 + mny2, mnx2 + mny1);
        }
        return mnx1 + mny1;
    }
};

int main(int argc, char *argv[]) {
    Solution test;
    cout << test.minOperations(temp_vector({1, 4, 2, 8}), 3) << endl;
    cout << test.minOperations(temp_vector({1, 1, 1}), 3) << endl;
    cout << test.minOperations(temp_vector({73, 92, 31, 78, 89}), 17) << endl;
}

//
// Created By jing At 2026-05-23 22:33
//
